examples/merkle_proofs.md

```diff
@@ -0,0 +1,157 @@
+This example demonstrates the _Merkle proof_ feature of Irmin on the task of producing tamper-proof banking account statements.
+
+From a high-level perspective, a Merkle proof is a Merkle tree where some subtrees are missing and replaced by their hash. A Merkle tree is simply a DAG where the hash of a node is derived from the hash of its children.
+
+More specifically, for Irmin, a Merkle proof is the subset of a node and contents tree stored in a repo.
+
+### Setting up a Bank
+
+```ocaml
+# #require "digestif.ocaml";;
+# #require "checkseum.ocaml";;
+# #require "irmin";;
+# #require "irmin-unix";;
+# #require "ppx_irmin";;
+# open Lwt.Infix
+  open Lwt.Syntax;;
+```
+
+First, create an irmin-unix store module which uses `int` as contents.
+
+```ocaml
+module Contents = struct
+  type t = int [@@deriving irmin]
+
+  let merge = Irmin.(Merge.idempotent Type.(option int))
+end
+
+module Store = Irmin_unix.Git.FS.KV (Contents)
+```
+
+Open a repo.
+
+```ocaml
+# let config = Irmin_git.config ~bare:true "/tmp/irmin/test"
+  let repo = Lwt_main.run (Store.Repo.v config);;
+val config : Irmin.config = <abstr>
+val repo : Store.repo = <abstr>
+```
+
+Create a tree which contains the accounts and their balance for 3 customers: Ben, Bob and Eve.
+
+Instead of using `[ "eve" ]` as a path (which whould have been valid too), this example splits the names char by char. This is better in order to highlight how proofs work.
+
+```ocaml
+# let tree = Lwt_main.run (
+    let tree = Store.Tree.empty () in
+    let* tree = Store.Tree.add tree [ "b"; "e"; "n" ] 10 in
+    let* tree = Store.Tree.add tree [ "b"; "o"; "b" ] 20 in
+    let+ tree = Store.Tree.add tree [ "e"; "v"; "e" ] 30 in
+    tree
+  );;
+val tree : Store.tree = <abstr>
+```
+
+In order to produce a Merkle proofs, Irmin requires that the tree on which the proof will be based is persisted inside a `repo`. The following code block does so.
+
+`tree_key` is a value that encodes where `tree` has been persisted inside the store's backend.
+
+```ocaml
+# let tree_key = Lwt_main.run (
+    (* [batch] exposes [repo] stores in read-write mode *)
+    let+ kinded_key = Store.Backend.Repo.batch repo
+      (fun rw_contents_store rw_node_store _rw_commit_store ->
+        Store.save_tree repo rw_contents_store rw_node_store tree)
+    in
+    (* As the root of [tree] is a node, refining the type is safe *)
+    match kinded_key with
+    | `Node key -> key
+    | `Contents _ -> assert false
+  );;
+val tree_key : Store.node_key = <abstr>
+```
+
+The bank is ready. `repo` contains the data, which was persisted on disk. `tree_key` acts as a pointer to the root of the accounts tree.
```

Comment:
The bank is ready. `repo` contains the data, which was persistent on disk. `tree_key` acts as a pointer to the root of the accounts tree.